src/coreclr/vm/peimagelayout.cpp

```diff
@@ -721,14 +741,60 @@ FlatImageLayout::FlatImageLayout(PEImage* pOwner)
         _ASSERTE((offset - mapBegin) < mapSize);
         _ASSERTE(mapSize >= (UINT64)size);
 
-        char *addr = (char*)CLRMapViewOfFile(m_FileMap, FILE_MAP_READ, mapBegin >> 32, (DWORD)mapBegin, (DWORD)mapSize);
-        if (addr == NULL)
+        LPVOID view = CLRMapViewOfFile(m_FileMap, FILE_MAP_READ, mapBegin >> 32, (DWORD)mapBegin, (DWORD)mapSize);
+        if (view == NULL)
             ThrowLastError();
 
-        addr += (offset - mapBegin);
-        m_FileView.Assign((LPVOID)addr);
+        m_FileView.Assign(view);
+        addr = (LPVOID)((size_t)view + offset - mapBegin);
+
+        INT64 uncompressedSize = pOwner->GetUncompressedSize();
+        if (uncompressedSize > 0)
+        {
+#if defined(CORECLR_EMBEDDED)
+            // The mapping we have just created refers to the region in the bundle that contains compressed data.
+            // We will create another anonymous memory-only mapping and uncompress file there.
+            // The flat image will refer to the anonimous mapping instead and we will release the original mapping.
+            HandleHolder anonMap = WszCreateFileMapping(INVALID_HANDLE_VALUE, NULL, PAGE_READWRITE, uncompressedSize >> 32, (DWORD)uncompressedSize, NULL);
+            if (anonMap == NULL)
+                ThrowLastError();
+
+            LPVOID anonView = CLRMapViewOfFile(anonMap, FILE_MAP_READ | FILE_MAP_WRITE, 0, 0, 0);
+            if (anonView == NULL)
+                ThrowLastError();
+
+            //NB: PE cannot be larger than 4GB and we are decompressing a managed assembly, which is a PE image,
+            //    thus converting sizes to uint32 is ok.
+            PAL_ZStream zStream;
+            zStream.nextIn = (uint8_t*)addr;
+            zStream.availIn = (uint32_t)size;
+            zStream.nextOut = (uint8_t*)anonView;
+            zStream.availOut = (uint32_t)uncompressedSize;
+
+            // we match the compression side here. 15 is the window sise, negative means no zlib header.
+            const int Deflate_DefaultWindowBits = -15;
+            if (CompressionNative_InflateInit2_(&zStream, Deflate_DefaultWindowBits) != PAL_Z_OK)
+                ThrowLastError();
+
+            int ret = CompressionNative_Inflate(&zStream, PAL_Z_NOFLUSH);
+            CompressionNative_InflateEnd(&zStream);
+            if (ret < 0)
+                ThrowLastError();
+
+            addr = anonView;
+            size = uncompressedSize;
+            // Replace file handles with the handles to anonymous map. This will release the handles to the original view and map.
+            m_FileView.Assign(anonView);
+            m_FileMap.Assign(anonMap);
+
+#else
+            _ASSERTE(!"Failure extracting contents of the application bundle. Compressed files used with a standalone (not singlefile) apphost.");
+            ThrowLastError();
```

Comment:
Will this actually fail correctly - `ThrowLastError` presumably throws the last Win32 error code, but in this case there was no OS call which failed before this... so the last error code is effectively random value.

Reply:
This is a failfast for "impossible" situation. Something is badly corrupted and we cannot recover, if we hit this. 
I will check what we use in such cases.